app/services/user_service.rb

```diff
@@ -4,9 +4,27 @@ def self.delete_spam(user)
     user.destroy
   end
 
+  def self.deactivate(user:, actor:, params:)
+    actor.ability.authorize! :deactivate, user
+    UserDeactivationResponse.create(user: user, body: params[:deactivation_response])
+    user.deactivate!
+    user
+  end
+
   def self.update(user:, actor:, params:)
     actor.ability.authorize! :update, user
-    user.update! params
+    user.update params
+    user
+  end
+
+  def self.change_password(user:, actor:, params:)
+    actor.ability.authorize! :update, user
+    user.errors.add :current_password, I18n.t(:"error.current_password_did_not_match") unless user.valid_password? params.delete(:current_password)
```

Comment:
oh are they the controller tests rather than service specs?

On Fri, 3 Jul 2015 at 08:18 James Kiesel <notifications@github.com> wrote:

> In app/services/user_service.rb
> <https://github.com/loomio/loomio/pull/2282#discussion_r33818276>:
>
> >    def self.update(user:, actor:, params:)
> >      actor.ability.authorize! :update, user
> > -    user.update! params
> > +    user.update params
> > +    user
> > +  end
> > +
> > +  def self.change_password(user:, actor:, params:)
> > +    actor.ability.authorize! :update, user
> > +    user.errors.add :current_password, I18n.t(:"error.current_password_did_not_match") unless user.valid_password? params.delete(:current_password)
>
> Yeah, I have change password covered with
>
>    - It doesn't work when the current password isn't provided correctly
>    - It doesn't work when the password doesn't match the password
>    confirmation
>    - It works when both of those things are correct
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/loomio/loomio/pull/2282/files#r33818276>.
>